File: 25. Reverse Nodes in k-Group/reverseKGroup.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
    	val = x;
    }
}

public class reverseKGroup {
    public ListNode solution(ListNode head, int k) {
        int count = 0;
        int countReverse = 0;
    	ListNode check = head;
        ListNode curr = head;
    	ListNode dummy = null;
        while(count < k && check != null){
            check = check.next;
            count = count + 1;
        }
        System.out.println(head);
        if (count == k){

            ListNode nextTemp = null;
            while(countReverse < k && curr != null){
                nextTemp = curr.next;
                curr.next = dummy;
                dummy = curr;
                curr = nextTemp;
                countReverse = countReverse + 1;
            }
            if (nextTemp != null){
                head.next = solution(nextTemp, k);
            }
            return dummy;
        }
        else {
            return head;
        }
    	
    }

    public static void main(String[] argv){
        ListNode node = new ListNode(1);
        ListNode node1 = new ListNode(2);
        ListNode node2 = new ListNode(3);
        ListNode node3 = new ListNode(4);
        ListNode node4 = new ListNode(5);

        node.next = node1;
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;

        ListNode head = node;
        while(node != null){
            System.out.print(node.val + "->");
            node = node.next;
        }
        System.out.println(" ");

        int k = 2;
        reverseKGroup obj = new reverseKGroup();
        ListNode res = obj.solution(head, k);
        System.out.println(res);
        while(res != null){
            System.out.print(res.val + "->");
            res = res.next;
        }
        System.out.println(" ");
    }
}

File: 25. Reverse Nodes in k-Group/reverseKGroup.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class reverseKGroup:
	def solution(self, head, k):
		count = 0
		countReverse = 0
		check = head
		while count < k and check != None:
			count = count + 1
			check = check.next
		if count == k:
			curr, dummy = head, None
			while (countReverse < k and curr):
				curr, dummy, dummy.next = curr.next, curr, dummy
				countReverse = countReverse + 1
			if curr != None:
				head.next = self.solution(curr, k)
			return dummy
		else:
			return head

node = ListNode(1)
node1 = ListNode(2)
node2 = ListNode(3)
node3 = ListNode(4)
node4 = ListNode(5)

node.next = node1
node1.next = node2
node2.next = node3
node3.next = node4

head = node
str1 = ""
while head:
	str1 = str1 + str(head.val) + "->"
	head = head.next
print(str1)

obj = reverseKGroup()
k = 3
b = obj.solution(node, k)
head1 = b
str2 = ""
while head1:
	str2 = str2 + str(head1.val) + "->"
	head1 = head1.next
print(str2)